Reunion undocked clipboard API spec (internal team review)

metathinker opened the pull request:
This is a API spec proposal for the clipboard API that will be part of the Project Reunion SDK.

The API is exactly the same as UWP Clipboard within [Windows.ApplicationModel.DataTransfer](
https://docs.microsoft.com/uwp/api/windows.applicationmodel.datatransfer), except that:
1. the namespace is different
1. [Clipboard.Flush()](https://docs.microsoft.com/uwp/api/windows.applicationmodel.datatransfer.clipboard.flush?view=winrt-19041) is now completely optional
1. DataPackage's Share-related events are omitted

This API spec is a draft in progress, and is published for the internal review of the clipboard team within Microsoft before more general publication in the main Project Reunion git repo. Folks outside Microsoft are of course welcome to comment as well, but if you do comment as a non-MSFT person, I may not be able to respond to your concern.

metathinker committed:
start of clipboard api v1 spec (motivation, copied examples)

metathinker committed:
completed draft

metathinker committed:
very minor

metathinker committed:
minor: add links, reformat markdown source

usvoyager commented on specs/clipboard-api/clipboard-api.md:
```diff
@@ -0,0 +1,312 @@
+clipboard API in the Reunion SDK will let us make future improvements
+to the clipboard developer and user experiences, like these:
+
+* **Improve clipboard reliability by removing traps in the clipboard APIs:**
```
should also address the foreground/background access restrictions here, what is the behavior these new apis going to support

usvoyager commented on specs/clipboard-api/clipboard-api.md:
```diff
@@ -0,0 +1,312 @@
+SDK clipboard API is exactly the same as Windows.ApplicationModel.DataTransfer,
+except for these differences:
+
+1. The namespace is now Microsoft.ProjectReunion.ApplicationModel.DataTransfer.
```
>DataTransfer [](start = 66, length = 12)

lets drop DataTransfer from the namespace, it doesn't add any value or context

metathinker committed:
rename namespace; add background clipboard access as a motivation

metathinker committed:
create clipboard.idl - reformatted copy of windows.applicationmodel.datatransfer.idl from SDK

metathinker committed:
s/Windows.Foundation.UniversalApiContract/ClipboardContract/g

metathinker committed:
change all GUIDs and temporarily rename all enums; now builds clean

metathinker committed:
rewrite Clipboard in MIDL3; collapse IClipboardStatics and IClipboardStatics2 into 1 interface

metathinker committed:
convert other all-statics classes to MIDL3 and consolidate their interfaces

metathinker committed:
convert Clipboard helper classes to MIDL3

metathinker committed:
convert DataPackage to MIDL3 and combine interfaces

metathinker committed:
convert DataPackageView to MIDL3 and combine interfaces

metathinker committed:
convert DataPackagePropertySet, DataPackagePropertySetView to MIDL3 and combine interfaces

metathinker committed:
Brought back separate IDataPackagePropertySet, IDataPackagePropertySetView

Both exist so they can `require` their implementors to also implement
IMap/IMapView<String, Object> and IIterable<IKeyValuePair<String, Object>>.

In the corresponding IDL from the 20H1/Vb Windows SDK, only
IDataPackagePropertySet does have the `require`. However, from what I
can tell, IDataPackagePropertySetView also had it in releases before
1809/RS5, and it looks like the `require` was mistakenly omitted when we
converted the internal IDL corresponding to the SDK IDL from MIDLRT to MIDL3.

metathinker committed:
Converted remaining Clipboard-relevant types to MIDL3

DataProviderRequest
DataProviderDeferral
DataProviderHandler

metathinker committed:
remove all Share-related types and members

DataPackage.ShareCompleted
DataPackage.ShareCanceled

DataTransferManager
DataRequest
DataRequestDeferral
DataRequestedEventArgs
ShareCompletedEventArgs
ShareProvider
ShareProviderHandler
ShareProviderOperation
ShareProvidersRequestedEventArgs
ShareTargetInfo
ShareUITheme
ShareUIOptions
SharedStorageAccessManager
TargetApplicationChosenEventArgs

metathinker committed:
change all contract versions to 1.0

metathinker committed:
leading tabs to spaces

metathinker committed:
javadoc; link from spec doc; copy paste bug fix

usvoyager commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,450 @@
+  static runtimeclass StandardDataFormats
+  {
+    static String Text { get; };
+    [deprecated("Uri may be altered or unavailable in future releases. Instead, use WebLink or ApplicationLink.", deprecate, ClipboardContract, 1.0)]
```
>deprecated [](start = 5, length = 10)

Please remove all deprecated, they are deprecated in windows sdk already no need to carry forward

metathinker committed:
remove [deprecated] Uri from DataPackage, StandardDataFormats

metathinker committed:
remove namespace prefix where not needed

usvoyager commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,445 @@
+
+  typedef enum ClipboardHistoryItemsResultStatus ClipboardHistoryItemsResultStatus;
+  typedef enum DataPackageOperation DataPackageOperation;
+  typedef enum SetHistoryItemAsContentStatus SetHistoryItemAsContentStatus;
```
this is old style MIDL

usvoyager commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,445 @@
+  /// includes clipboard items roamed from your other devices through the cloud.
+  /// </summary>
+  [contract(ClipboardContract, 1.0)]
+  [static_name("IClipboardStatics", 6cb53030-4da0-43a1-95ab-52f87cf59c3a)]
```
>[static_name("IClipboardStatics", 6cb53030-4da0-43a1-95ab-52f87cf59c3a)] [](start = 2, length = 72)

remove all custom uuids, cppwinrt will generate it

static_name, interface_name

metathinker replied:
normally I'm not a fan of letting midlrt.exe generate the GUIDs, it makes it impossible to search for them in the original source code (.idl) itself and you have to generate the .h first.

but it's not a big deal, so I removed them.

usvoyager commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,445 @@
+  // TO CONSIDER: should this class be omitted since it was empty in-box?
+  [contract(ClipboardContract, 1.0)]
+  [interface_name("IClipboardHistoryChangedEventArgs", 9d1960c7-cf0a-4e51-943c-72373b7a2579)]
+  runtimeclass ClipboardHistoryChangedEventArgs
```
>ClipboardHistoryChangedEventArgs [](start = 15, length = 32)

lets omit empty classes

metathinker committed:
respond to CR feedback

- remove the empty class ClipboardHistoryChangedEventArgs;
  change event Clipboard.HistoryChanged to take a generic
  IInspectable as the event handler input arg

- remove all explicitly set interface GUIDs;
  let midlrt.exe generate IIDs instead

- remove `typedef enum Blah Blah` as that is not needed in MIDL 3

metathinker committed:
reorder some members to follow more logical groupings rather than 'what version of modern/UWP it first appeared in'

metathinker committed:
minor

metathinker commented on specs/clipboard-api/clipboard-api.md:
```diff
@@ -0,0 +1,333 @@
+clipboard API in the [Windows.ApplicationModel.DataTransfer](
+https://docs.microsoft.com/uwp/api/windows.applicationmodel.datatransfer)
+WinRT namespace.
+In fact, in this initial version (as of mid-2020), the Reunion
+SDK clipboard API is the same as Windows.ApplicationModel.DataTransfer,
```
This does mean that we import the asynchronous model of most UWP classes, like the ones in Windows.ApplicationModel.DataTransfer. @usvoyager suggested at one point we should review the async behavior.

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+    static String CreateHtmlFormat(String htmlFragment);
+  }
+
+  /// <summary>
+  /// Input parameters for <see cref="DataPackage.OperationCompleted"/> event handler functions.
+  /// </summary>
+  [contract(ClipboardContract, 1.0)]
+  runtimeclass OperationCompletedEventArgs
+  // FIXME: since the namespace is now ApplicationModel, the name OperationCompletedEventArgs
+  // is too generic. Should we rename it?
```
as the FIXME comment says, this class should probably be renamed. Perhaps `DataPackageOperationCompletedEventArgs`? Or perhaps we should rethink `DataPackage.OperationCompleted` and the other event in `DataPackage` entirely.

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+    String Description;
+    Windows.Storage.Streams.IRandomAccessStreamReference Thumbnail;
+    Windows.Foundation.Collections.IVector<String> FileTypes { get; };
+    String ApplicationName;
+    Windows.Foundation.Uri ApplicationListingUri;
```
These DataPackagePropertySet properties assume that the DataPackage comes from an AppX/MSIX-packaged app. That was almost always true when this API was in UWP. Do we want to encode that assumption in the undocked version?

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+
+    Windows.Foundation.Uri ContentSourceWebLink;
+    Windows.Foundation.Uri ContentSourceApplicationLink;
+    String PackageFamilyName;
+    Windows.Storage.Streams.IRandomAccessStreamReference Square30x30Logo;
+    Windows.UI.Color LogoBackgroundColor;
```
(same as lines 311-312) These DataPackagePropertySet properties assume that the DataPackage comes from an AppX/MSIX-packaged app. That was almost always true when this API was in UWP. Do we want to encode that assumption in the undocked version?

metathinker committed:
clarifications per chat with Ieva

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+    DataPackageOperation RequestedOperation;
+
+    event Windows.Foundation.TypedEventHandler<DataPackage, OperationCompletedEventArgs> OperationCompleted;
+    event Windows.Foundation.TypedEventHandler<DataPackage, Object> Destroyed;
```
We should probably delete this event. The [documentation for DataPackage.Destroyed](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.datatransfer.datapackage.destroyed?view=winrt-19041) says "In general, the system handles the destruction of a DataPackage object. Your app should not have to handle this event." I agree as I can't imagine any use for it not already covered by OperationCompleted. Plus it confuses the strange ownership model of DataPackage (does the system own it? if so, why does the source app still need to care about its lifetime?) even more.

But before deleting the event, I should dig up the original Windows 8-era internal API specs to see why the event was added originally.

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+  {
+    DataPackageOperation Operation { get; };
+
+    String AcceptedFormatId { get; };
```
This property is really poorly named, and [the documentation](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.datatransfer.operationcompletedeventargs.acceptedformatid?view=winrt-19041) does little more than repeat the name. I should investigate what the property even does before retaining it.

My best guess before doing that investigation is that on a paste (Clipboard.GetContent() followed by DataPackageView.GetDataAsync(format) on the return value from Clipboard), the DataPackage.OperationCompleted event is signaled, with this property set to the name of the data format that the target app read out.

metathinker commented on specs/clipboard-api/clipboard.idl:
```diff
@@ -0,0 +1,426 @@
+    DataPackagePropertySet Properties { get; };
+    DataPackageOperation RequestedOperation;
+
+    event Windows.Foundation.TypedEventHandler<DataPackage, OperationCompletedEventArgs> OperationCompleted;
```
The documentation for this event DataPackage.OperationCompleted doesn't say when your app should handle the event. My best guess is that you handle it in order to know when you can delete any resources associated with a delay-rendered format. (If true, this gives the OperationCompletedEventArgs.AcceptedFormatId property some use, as you need to know what format was read in order to know whether you delay-rendered it or not.)

metathinker committed:
minor updates per spec review with Ieva

metathinker commented:
closing in favor of upcoming (more) public review.
